OpenRA.Mods.D2k/Traits/Sandworm.cs

```diff
@@ -0,0 +1,110 @@
+#region Copyright & License Information
+/*
+* Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+* This file is part of OpenRA, which is free software. It is made
+* available to you under the terms of the GNU General Public License
+* as published by the Free Software Foundation. For more information,
+* see COPYING.
+*/
+#endregion
+
+using System;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.D2k.Traits
+{
+	class SandwormInfo : WandersInfo
+	{
+		public const string IdleSequenceName = "idle";
+		public const int TargetRescanIntervalInTicks = 5;
+
+		public override object Create(ActorInitializer init) { return new Sandworm(init.Self, this); }
+	}
+
+	class Sandworm : Wanders, ITick, INotifyBecomingIdle
+	{
+		readonly Lazy<IMove> move;
+		readonly Lazy<RenderUnit> renderUnit;
+		readonly Lazy<AttackBase> attackTrait;
+		readonly Lazy<ReducesNoise> noiseReduction;
+
+		int targetCountdown;
+		uint currentTargetActorId;
+
+		public Sandworm(Actor self, SandwormInfo info)
+			: base(self, info)
+		{
+			move = Exts.Lazy(self.Trait<IMove>);
+			renderUnit = Exts.Lazy(self.Trait<RenderUnit>);
+			attackTrait = Exts.Lazy(self.Trait<AttackBase>);
+			noiseReduction = Exts.Lazy(() => self.World.WorldActor.Trait<ReducesNoise>());
+		}
+
+		public new void OnBecomingIdle(Actor self)
+		{
+			if (renderUnit.Value.DefaultAnimation.CurrentSequence.Name != SandwormInfo.IdleSequenceName)
+				renderUnit.Value.DefaultAnimation.ReplaceAnim("idle");
+
+			base.OnBecomingIdle(self);
+		}
+
+		public override void DoAction(Actor self, CPos targetCell)
+		{
+			Game.Debug("(Sandworm {0}) is wandering.", self.ActorID);
+			self.QueueActivity(move.Value.MoveWithinRange(Target.FromCell(self.World, targetCell, SubCell.Any), WRange.FromCells(1)));
+		}
+
+		public void Tick(Actor self)
+		{
+			if (--targetCountdown > 0)
+				return;
+
+			targetCountdown = SandwormInfo.TargetRescanIntervalInTicks;
+
+			var food = self.World.ActorsWithTrait<MakesNoise>();
+			var desiredPair = new TraitPair<MakesNoise>();
+
+			// Basically any food would do
+			var maxStrength = int.MinValue;
+
+			foreach (var traitPair in food)
```

Comment:
The way I would do this:

1. Move `NoiseAttenuation` into `MakesNoiseInfo` and drop the unnecessary `Noise` prefix from that and `NoiseIntensity`.
2. Remove the `DangerousTerrain` property - we don't have a use case for units generating noise on terrain types that the worms can't travel on.
3. Define a `MaxSearchRadius` property on `SandwormInfo` with a sensible value, probably 10 - 20 cells.  Worms shouldn't care about things that are happening on the other side of the map.
4. Loop over `world.FindActorsInCircle(MaxSearchRadius)`.
  - If the actor doesn't generate noise or is on unpathable terrain, then skip it.
  - If `wormLocation == actorLocation`, then attack the current cell and abort further evaluation.
  - If not, then accumulate the contribution of the actor in a "direction of most noise" vector: `noise += actorNoise * (wormPosition - actorPosition) / actorAttenuation;`
5. If the noise vector is non-zero then move in that direction, otherwise move in a random direction.